### PR TITLE
Enforce stricter validation for data read from binary file

### PR DESCRIPTION
**Pull request by jiangliu**
Enforce stricter validation for data read from binary file by:
- limit maximum size of binary file
- validate slice indices
- use checked_add to avoid overflow
- skip imported symbol
- skip symbols with values out of range

### CODE REVIEW

**Review by benfred — COMMENTED** [review]
thanks for the PR!

Can you say a little bit about the problem you're having that this code is fixing?

File: src/utils.rs [hunk 1]
```diff
@@ -20,3 +20,60 @@ pub fn resolve_filename(filename: &str, modulename: &str) -> Option<String> {
...
+pub fn is_subrange_usize(start: usize, size: usize, sub_start: usize, sub_size: usize) -> bool {
+    size != 0
+        && sub_size != 0
+        && start.checked_add(size).is_some()
+        && sub_start.checked_add(sub_size).is_some()
+        && sub_start >= start
+        && sub_start + sub_size <= start + size
+}
+
+pub fn is_subrange_u64(start: u64, size: u64, sub_start: u64, sub_size: u64) -> bool {
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/utils.rs:33]: Yeah, I have tried to implement a generic function, but found we need to introduce an extra dependency on `num_traits`. If it's ok to depend on `num_traits`, I will give it a try:)

**Comment by jiangliu** [comment]
> thanks for the PR!
> 
> Can you say a little bit about the problem you're having that this code is fixing?

We are trying to use py-spy to dump stack trace of current process/thread instead of remote process, so we hope to harden py-spy to avoid invalid memory access. It would also help to protect py-spy in codedump mode when the coredump files are from untrusted sources:)

File: src/binary_parser.rs [hunk 2]
```diff
@@ -6,6 +6,10 @@ use anyhow::Error;
 use goblin::Object;
 use memmap::Mmap;
 
+use crate::utils::{is_subrange_u64, is_subrange_usize};
+
+const MAX_BINARY_FILE_SIZE: u64 = 0x80000000;
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/binary_parser.rs:11]: Not really, it's just defensive:)
How about enlarge the limitation? Or just remove it?

[the remark this answers is not in the document]
Comment (benfred) [reply src/binary_parser.rs:11]: If this isn't solving a problem you've hit before - lets just remove the filesize limitation . 

The rest of the changes look good btw!

[the remark this answers is not in the document]
Comment (jiangliu) [reply src/binary_parser.rs]: Remove the file size check.

**Pull request merged**